modules/table/src/main/java/org/apache/ignite/internal/table/LiveSchemaTupleBuilderImpl.java

```diff
@@ -0,0 +1,133 @@
+ */
+
+package org.apache.ignite.internal.table;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import org.apache.ignite.internal.schema.Column;
+import org.apache.ignite.internal.schema.SchemaRegistry;
+import org.apache.ignite.internal.schema.marshaller.MarshallerUtil;
+import org.apache.ignite.internal.table.distributed.TableManager;
+import org.apache.ignite.schema.ColumnType;
+import org.apache.ignite.schema.SchemaBuilders;
+import org.apache.ignite.table.Tuple;
+import org.apache.ignite.table.TupleBuilder;
+
+import static org.apache.ignite.internal.schema.configuration.SchemaConfigurationConverter.convert;
+
+/**
+ * Live schema buildable tuple.
+ *
+ * Allows to create columns implicitly by adding previously nonexistent columns during insert.
+ */
+public class LiveSchemaTupleBuilderImpl extends TupleBuilderImpl {
+    /** Live schema column values. */
+    private final Map<String, Object> liveSchemaColMap;
+
+    /** Schema registry. */
+    private final SchemaRegistry schemaRegistry;
+
+    /** Current table name. */
+    private final String tblName;
+
+    /** Table manager. */
+    private final TableManager mgr;
+
+    /**
+     * Constructor.
+     */
+    public LiveSchemaTupleBuilderImpl(SchemaRegistry schemaRegistry, String tblName, TableManager mgr) {
+        super(schemaRegistry == null ? null : schemaRegistry.schema());
+
+        Objects.requireNonNull(schemaRegistry);
+        Objects.requireNonNull(tblName);
+        Objects.requireNonNull(mgr);
+
+        this.schemaRegistry = schemaRegistry;
+        this.tblName = tblName;
+        this.mgr = mgr;
+
+        liveSchemaColMap = new HashMap<>();
+    }
+
+    /** {@inheritDoc} */
+    @Override public TupleBuilder set(String colName, Object val) {
+        Column col = schema().column(colName);
+
+        if (col == null) {
+            if (val == null)
+                return this;
+
+            liveSchemaColMap.put(colName, val);
+            return this;
+        }
+        super.set(colName, val);
+
+        return this;
+    }
+
+    /** {@inheritDoc} */
+    @Override public Tuple build() {
+        Map<String, ColumnType> colTypeMap = new HashMap<>();
+
+        for (Map.Entry<String, Object> entry : liveSchemaColMap.entrySet()) {
+            String colName = entry.getKey();
+            Object val = entry.getValue();
+
+            ColumnType type = MarshallerUtil.columnType(val.getClass());
+
+            if (type == null)
+                throw new UnsupportedOperationException("Live schema update for type [" + val.getClass() + "] is not supported yet.");
+
+            colTypeMap.put(colName, type);
+        }
+
+        if (!colTypeMap.isEmpty()) {
+            createColumns(colTypeMap);
+            this.schema(schemaRegistry.schema());
+            rebuildTupleWithNewSchema();
+        }
+
+        liveSchemaColMap.forEach(super::set);
+
+        return this;
```

Comment:
As the method 'build()' should return valid Tuple, 
you can simply rewrite this in a next way returning a new Tuple instance:
```
if (colTypeMap.isEmpty())
    return this;

createColumns(colTypeMap);

return rebuildTupleWithNewSchema();
```
It looks ok to insert a row of a-bit-old-version in case of concurrent schema changing.
The result will be as if the row was added just before the foreign schema update and the row will be upgraded in the background anyway.
The only we should bother is the schema fit the data.